File: app/src/test/java/com/example/anton/sprintcalendar/SprintCalendarTest.java
package com.example.anton.sprintcalendar;

import org.joda.time.LocalDate;
import org.joda.time.format.DateTimeFormat;
import org.joda.time.format.DateTimeFormatter;
import org.junit.Test;

import java.text.ParseException;

import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.assertThat;

public class SprintCalendarTest {

    public static final String TODAY_DATE = "12.01.2016";
    private DateTimeFormatter formatter = DateTimeFormat.forPattern("dd.MM.yyyy");

    @Test
    public void populatesDatesOfTenDaysOfSprint() {

        SprintCalendar sprintCalendar = new SprintCalendar(
                new Team(), new TestDateProvider(TODAY_DATE),
                new DefaultHolidayProvider(new LocalDate(2016, 01, 06))
        );

        sprintCalendar.initByCurrentDate();

        assertThat(sprintCalendar.day(0).getDate(), is(date("04.01.2016")));
        assertThat(sprintCalendar.day(1).getDate(), is(date("05.01.2016")));
        assertThat(sprintCalendar.day(2).getDate(), is(date("06.01.2016")));
        assertThat(sprintCalendar.day(3).getDate(), is(date("07.01.2016")));
        assertThat(sprintCalendar.day(4).getDate(), is(date("08.01.2016")));
        assertThat(sprintCalendar.day(5).getDate(), is(date("11.01.2016")));
        assertThat(sprintCalendar.day(6).getDate(), is(date("12.01.2016")));
        assertThat(sprintCalendar.day(7).getDate(), is(date("13.01.2016")));
        assertThat(sprintCalendar.day(8).getDate(), is(date("14.01.2016")));
        assertThat(sprintCalendar.day(9).getDate(), is(date("15.01.2016")));
    }

    @Test
    public void determinesTodaySprintDay() throws ParseException {
        SprintCalendar sprintCalendar = new SprintCalendar(
                new Team(), new TestDateProvider("12.01.2016"),
                new DefaultHolidayProvider(new LocalDate(2016, 01, 06))
        );

        sprintCalendar.initByCurrentDate();

        assertThat(sprintCalendar.day(6).isToday(), is(true));
        assertThat(sprintCalendar.day(5).isToday(), is(false));
    }

    @Test
    public void determinesHoliday() {
        SprintCalendar sprintCalendar = new SprintCalendar(
                new Team(), new TestDateProvider("12.01.2016"),
                new DefaultHolidayProvider(new LocalDate(2016, 01, 06))
        );

        sprintCalendar.initByCurrentDate();

        assertThat(sprintCalendar.day(2).isHoliday(), is(true));
    }

    @Test
    public void calculatesLeftSprintDays() {
        SprintCalendar sprintCalendar = new SprintCalendar(
                new Team(), new TestDateProvider("12.01.2016"),
                new DefaultHolidayProvider()
        );

        sprintCalendar.initByCurrentDate();

        assertThat(sprintCalendar.getDaysLeft(), is(4));
    }

    @Test
    public void calculatesLeftSprintDaysIgnoringWeekEnds() {
        SprintCalendar sprintCalendar = new SprintCalendar(
                new Team(), new TestDateProvider("06.01.2016"),
                new DefaultHolidayProvider()
        );

        sprintCalendar.initByCurrentDate();

        assertThat(sprintCalendar.getDaysLeft(), is(8));
    }

    @Test
    public void calculatesLeftSprintDaysIgnoringHolidays() {
        SprintCalendar sprintCalendar = new SprintCalendar(
                new Team(), new TestDateProvider("12.01.2016"),
                new DefaultHolidayProvider(new LocalDate(2016, 01, 14))
        );

        sprintCalendar.initByCurrentDate();

        assertThat(sprintCalendar.getDaysLeft(), is(3));
    }

    @Test
    public void calculatesTotalSprintDaysIgnoringHolidays() {
        SprintCalendar sprintCalendar = new SprintCalendar(
                new Team(), new TestDateProvider("12.01.2016"),
                new DefaultHolidayProvider(new LocalDate(2016, 01, 06))
        );

        sprintCalendar.initByCurrentDate();

        assertThat(sprintCalendar.getTotalDays(), is(9));
    }

    @Test
    public void calculatesTotalSprintHours() {
        SprintCalendar sprintCalendar = new SprintCalendar(
                new Team(new TeamMember("John"), new TeamMember("Peter"), new TeamMember("Pedro")),
                new TestDateProvider("12.01.2016"),
                new DefaultHolidayProvider(new LocalDate(2016, 01, 06))
        );

        sprintCalendar.initByCurrentDate();

        assertThat(sprintCalendar.getTotalHours(), is(9 * 3 * 5));
    }

    @Test
    public void calculatesTotalSprintHoursExcludingAbsenceOfTeamMembers() {
        TeamMember peter = new TeamMember("Peter");
        peter.setPresence(new LocalDate(2016, 01, 14), PresenceType.NONE);
        SprintCalendar sprintCalendar = new SprintCalendar(
                new Team(new TeamMember("John"), peter, new TeamMember("Pedro")),
                new TestDateProvider("12.01.2016"),
                new DefaultHolidayProvider(new LocalDate(2016, 01, 06))
        );

        sprintCalendar.initByCurrentDate();

        assertThat(sprintCalendar.getTotalHours(), is(9 * 3 * 5 - 5));
    }

    @Test
    public void calculatesLeftSprintHours() {
        SprintCalendar sprintCalendar = new SprintCalendar(
                new Team(new TeamMember("John"), new TeamMember("Peter"), new TeamMember("Pedro")),
                new TestDateProvider("12.01.2016"),
                new DefaultHolidayProvider(new LocalDate(2016, 01, 06))
        );

        sprintCalendar.initByCurrentDate();

        assertThat(sprintCalendar.getHoursLeft(), is(4 * 3 * 5));
    }

    @Test
    public void calculatesLeftSprintHoursExcludingAbsenceOfTeamMembers() {
        TeamMember peter = new TeamMember("Peter");
        peter.setPresence(new LocalDate(2016, 01, 14), PresenceType.NONE);
        SprintCalendar sprintCalendar = new SprintCalendar(
                new Team(new TeamMember("John"), peter, new TeamMember("Pedro")),
                new TestDateProvider("12.01.2016"),
                new DefaultHolidayProvider(new LocalDate(2016, 01, 06))
        );

        sprintCalendar.initByCurrentDate();

        assertThat(sprintCalendar.getHoursLeft(), is(4 * 3 * 5 - 5));
    }

    private LocalDate date(String dateString) {
        return formatter.parseDateTime(dateString).toLocalDate();
    }

    private class TestDateProvider implements DateProvider {

        private String todayDate;

        private TestDateProvider(String todayDate) {
            this.todayDate = todayDate;
        }

        @Override
        public boolean isToday(LocalDate date) {
            return date(todayDate).compareTo(date) == 0;
        }

        @Override
        public LocalDate getToday() {
            return date(todayDate);
        }
    }

}

File: app/src/test/java/net/sourcefusion/agiletools/sprintcalendar/SprintCalendarTest.kt
package net.sourcefusion.agiletools.sprintcalendar

import net.sourcefusion.agiletools.sprintcalendar.persisting.stubs.StubTeamRepository
import org.hamcrest.CoreMatchers.`is`
import org.joda.time.LocalDate
import org.joda.time.LocalTime
import org.joda.time.format.DateTimeFormat
import org.junit.Assert.assertThat
import org.junit.Test
import java.text.ParseException

class SprintCalendarTest {
    private val formatter = DateTimeFormat.forPattern("dd.MM.yyyy")

    @Test
    fun populatesDatesOfTenDaysOfSprint() {

        val sprintCalendar = SprintCalendar(
                StubTeamRepository(Team()), TestDateProvider(date(TODAY_DATE), LocalTime(9, 0)),
                BasicHolidayProvider(LocalDate(2016, 1, 6)))

        sprintCalendar.initByCurrentDate()

        assertThat(sprintCalendar.day(0).date, `is`("04.01.2016".toLocalDate()))
        assertThat(sprintCalendar.day(1).date, `is`("05.01.2016".toLocalDate()))
        assertThat(sprintCalendar.day(2).date, `is`("06.01.2016".toLocalDate()))
        assertThat(sprintCalendar.day(3).date, `is`("07.01.2016".toLocalDate()))
        assertThat(sprintCalendar.day(4).date, `is`("08.01.2016".toLocalDate()))
        assertThat(sprintCalendar.day(5).date, `is`("11.01.2016".toLocalDate()))
        assertThat(sprintCalendar.day(6).date, `is`("12.01.2016".toLocalDate()))
        assertThat(sprintCalendar.day(7).date, `is`("13.01.2016".toLocalDate()))
        assertThat(sprintCalendar.day(8).date, `is`("14.01.2016".toLocalDate()))
        assertThat(sprintCalendar.day(9).date, `is`("15.01.2016".toLocalDate()))
    }

    @Test
    @Throws(ParseException::class)
    fun determinesTodaySprintDay() {
        val sprintCalendar = SprintCalendar(
                StubTeamRepository(Team()), TestDateProvider("12.01.2016".toLocalDate(), LocalTime(9, 0)),
                BasicHolidayProvider(LocalDate(2016, 1, 6)))

        sprintCalendar.initByCurrentDate()

        assertThat(sprintCalendar.day(6).isToday, `is`(true))
        assertThat(sprintCalendar.day(5).isToday, `is`(false))
    }

    @Test
    fun determinesHoliday() {
        val sprintCalendar = SprintCalendar(
                StubTeamRepository(Team()), TestDateProvider("12.01.2016".toLocalDate(), LocalTime(9, 0)),
                BasicHolidayProvider(LocalDate(2016, 1, 6)))

        sprintCalendar.initByCurrentDate()

        assertThat(sprintCalendar.day(2).isHoliday, `is`(true))
    }

    @Test
    fun calculatesLeftSprintDays() {
        val sprintCalendar = SprintCalendar(
                StubTeamRepository(Team()), TestDateProvider("12.01.2016".toLocalDate(), LocalTime(9, 0)),
                BasicHolidayProvider())

        sprintCalendar.initByCurrentDate()

        assertThat(sprintCalendar.daysLeft, `is`(4))
    }

    @Test
    fun calculatesLeftSprintDaysIgnoringWeekEnds() {
        val sprintCalendar = SprintCalendar(
                StubTeamRepository(Team()), TestDateProvider("06.01.2016".toLocalDate(), LocalTime(9, 0)),
                BasicHolidayProvider())

        sprintCalendar.initByCurrentDate()

        assertThat(sprintCalendar.daysLeft, `is`(8))
    }

    @Test
    fun calculatesLeftSprintDaysIgnoringHolidays() {
        val sprintCalendar = SprintCalendar(
                StubTeamRepository(Team()), TestDateProvider("12.01.2016".toLocalDate(), LocalTime(9, 0)),
                BasicHolidayProvider(LocalDate(2016, 1, 14)))

        sprintCalendar.initByCurrentDate()

        assertThat(sprintCalendar.daysLeft, `is`(3))
    }

    @Test
    fun calculatesTotalSprintDaysIgnoringHolidays() {
        val sprintCalendar = SprintCalendar(
                StubTeamRepository(Team()), TestDateProvider("12.01.2016".toLocalDate(), LocalTime(9, 0)),
                BasicHolidayProvider(LocalDate(2016, 1, 6)))

        sprintCalendar.initByCurrentDate()

        assertThat(sprintCalendar.totalDays, `is`(9))
    }

    @Test
    fun calculatesTotalSprintHours() {
        val sprintCalendar = SprintCalendar(
                StubTeamRepository(Team(TeamMember("John"), TeamMember("Peter"), TeamMember("Pedro"))),
                TestDateProvider("12.01.2016".toLocalDate(), LocalTime(9, 0)),
                BasicHolidayProvider(LocalDate(2016, 1, 6)))

        sprintCalendar.initByCurrentDate()

        assertThat(sprintCalendar.totalHours, `is`(9 * 3 * 5))
    }

    @Test
    fun calculatesTotalSprintHoursExcludingAbsenceOfTeamMembers() {
        val peter = TeamMember("Peter")
        peter.setPresence(LocalDate(2016, 1, 14), PresenceType.ABSENT)
        val sprintCalendar = SprintCalendar(
                StubTeamRepository(Team(TeamMember("John"), peter, TeamMember("Pedro"))),
                TestDateProvider("12.01.2016".toLocalDate(), LocalTime(9, 0)),
                BasicHolidayProvider(LocalDate(2016, 1, 6)))

        sprintCalendar.initByCurrentDate()

        assertThat(sprintCalendar.totalHours, `is`(9 * 3 * 5 - 5))
    }

    @Test
    fun calculatesLeftSprintHours() {
        val sprintCalendar = SprintCalendar(
                StubTeamRepository(Team(TeamMember("John"), TeamMember("Peter"), TeamMember("Pedro"))),
                TestDateProvider("12.01.2016".toLocalDate(), LocalTime(9, 0)),
                BasicHolidayProvider(LocalDate(2016, 1, 6)))

        sprintCalendar.initByCurrentDate()

        assertThat(sprintCalendar.hoursLeft, `is`(4 * 3 * 5))
    }

    @Test
    fun calculatesLeftSprintHoursExcludingAbsenceOfTeamMembers() {
        val peter = TeamMember("Peter")
        peter.setPresence(LocalDate(2016, 1, 14), PresenceType.ABSENT)
        val sprintCalendar = SprintCalendar(
                StubTeamRepository(Team(TeamMember("John"), peter, TeamMember("Pedro"))),
                TestDateProvider("12.01.2016".toLocalDate(), LocalTime(9, 0)),
                BasicHolidayProvider(LocalDate(2016, 1, 6)))

        sprintCalendar.initByCurrentDate()

        assertThat(sprintCalendar.hoursLeft, `is`(4 * 3 * 5 - 5))
    }

    @Test
    fun calculatesLeftSprintHoursExcludingPastHoursOfTheDay() {
        val sprintCalendar = SprintCalendar(
                StubTeamRepository(Team(TeamMember("John"), TeamMember("Peter"), TeamMember("Pedro"))),
                TestDateProvider("12.01.2016".toLocalDate(), LocalTime(11, 0)),
                BasicHolidayProvider(LocalDate(2016, 1, 6)))

        sprintCalendar.initByCurrentDate()

        assertThat(sprintCalendar.hoursLeft, `is`(4 * 3 * 5 - 4))
    }

    @Test
    fun calculatesLeftSprintHoursExcludingPastHoursOfTheDayOnlyForPresentMembers() {
        val peter = TeamMember("Peter")
        peter.setPresence(LocalDate(2016, 1, 12), PresenceType.ABSENT)

        val sprintCalendar = SprintCalendar(
                StubTeamRepository(Team(TeamMember("John"), peter, TeamMember("Pedro"))),
                TestDateProvider("12.01.2016".toLocalDate(), LocalTime(11, 0)),
                BasicHolidayProvider(LocalDate(2016, 1, 6)))

        sprintCalendar.initByCurrentDate()

        assertThat(sprintCalendar.hoursLeft, `is`(4 * 3 * 5 - 5 - 3))
    }

    @Test
    fun calculatesLeftSprintHoursExcludingMorningPastHoursOfTheDayOnlyForPresentMembersInTheMorning() {
        val peter = TeamMember("Peter")
        peter.setPresence(LocalDate(2016, 1, 12), PresenceType.HALF_DAY)
        val pedro = TeamMember("Pedro")
        pedro.setPresence(LocalDate(2016, 1, 12), PresenceType.HALF_DAY_MORNING)

        val sprintCalendar = SprintCalendar(
                StubTeamRepository(Team(TeamMember("John"), peter, pedro)),
                TestDateProvider("12.01.2016".toLocalDate(), LocalTime(11, 0)),
                BasicHolidayProvider(LocalDate(2016, 1, 6)))

        sprintCalendar.initByCurrentDate()

        assertThat(sprintCalendar.hoursLeft, `is`(4 * 3 * 5 - 5 - 3))
    }

    @Test
    fun calculatesLeftSprintHoursExcludingEveningPastHoursOfTheDayOnlyForPresentMembersInTheEvening() {
        val peter = TeamMember("Peter")
        peter.setPresence(LocalDate(2016, 1, 12), PresenceType.HALF_DAY)
        val pedro = TeamMember("Pedro")
        pedro.setPresence(LocalDate(2016, 1, 12), PresenceType.HALF_DAY_MORNING)

        val sprintCalendar = SprintCalendar(
                StubTeamRepository(Team(TeamMember("John"), peter, pedro)),
                TestDateProvider("12.01.2016".toLocalDate(), LocalTime(15, 0)),
                BasicHolidayProvider(LocalDate(2016, 1, 6)))

        sprintCalendar.initByCurrentDate()

        assertThat(sprintCalendar.hoursLeft, `is`(4 * 3 * 5 - 13))
    }

    private fun date(dateString: String): LocalDate {
        return formatter.parseDateTime(dateString).toLocalDate()
    }

    companion object {
        val TODAY_DATE = "12.01.2016"
    }

}
